CORE-6140: FakeCorda creates fake virtual nodes

LizR3 opened the pull request:
Virtual nodes are associated with more than just a member. FakeCorda now works with a HoldingIdentity class and explicitly creates "virtual nodes" rather than just `upload`. Also takes multiple vararg flows on upload.

LizR3 committed:
CORE-6140: Cordapp test utils using `HoldingIdentity`

Virtual nodes are associated with more than just a member. FakeCorda now works with a HoldingIdentity class and explicitly creates "virtual nodes" rather than just `upload`. Also takes multiple vararg flows on upload.

LizR3 commented:
Hi @driessamyn, we talked about FakeCorda returning a `VirtualNode` and running the flows on that, but it felt wrong to be pulling out FakeCorda's internals (especially since all the VNs would have the same FakeFiber running through them) so I left the `invoke` method in place. Welcome your feedback before bugging all the usual suspects.

LizR3 committed:
CORE-6140: RPCRequestDataMock -> ...Wrapper

driessamyn commented on cordapp-test-utils/example-app/src/integrationTest/kotlin/net/cordacon/example/RollCallTest.kt:
```diff
@@ -13,25 +14,28 @@ class RollCallTest {
         val corda = FakeCorda()
-        val teacher = MemberX500Name.parse("CN=Ben Stein, OU=Economics, O=Glenbrook North High School, L=Chicago, C=US")
-        corda.upload(teacher, RollCallFlow::class.java)
+        val teacher = MemberX500Name.parse(
+            "CN=Ben Stein, OU=Economics, O=Glenbrook North High School, L=Chicago, C=US")
+        val teacherId = HoldingIdentity(teacher)
```
I think it'd be great to have shortcuts for these.
For example, maybe `val teacherId = HoldingIdentity.for("Ben Stein")` generates a valid X500 based on defaults (that can be overridden).
Taking this a step further, maybe you have something like:

```kotlin
// this creates a DTO for VNode metadata, generating things that can be default like full X500 and keys
val benVNode = VNode.for("Ben Stein", listOf(RollCallFlow::class.java, RollCallContract::class.java))
// not sure whether `deploy` is the right verb here, alternative could be `enable` or `configure`
corda.deploy(benVNode)
```

I see your argument about "leaking" test code, but vnode here is just metadata that can be interrogated as well (e.g `benVNode.id` or `benVNode.publicKey`).
So the invocation code below can still be something like:

```kotlin
corda.invokeFlow(benVNode.id, ...)
```

However, I think it may actually be more intuitive to do something like:

```kotlin
val network = FakeCorda()
val codeUnderTest = listOf(RollCallFlow::class.java, RollCallContract::class.java, AbsenceCallResponderFlow::class.java)
val teacherNode = VNode.for(network, "Ben Stein", codeUnderTest)
val studentNodes = listOf("Albers", "Anderson", "Anheiser", "Busch", "Bueller"). map {
            VNode.for(network, it, codeUnderTest|)
        }
teacherNode.startFlow(RPCRequestDataMock.fromData(
            "r1",
            RollCallFlow::class.java,
            RollCallInitiationRequest(studentNodes.map { it.id.x500 } )
)) 
```

I think this feels more natural because you always start a flow from a node. You can't go to the corda network and say "start a flow for X", you interact with the node X and say "start flow". 

The above example reflects the real API: 
 - initiate the flow from the Node
 - pass counterparty IDs in as an X500

But at the same time, all the counterparties (students) have registered to the network (implicitly) in the `VNode.for(...` operation, so Corda should be able to "find" them.

LizR3 replied:
`val teacherId = HoldingIdentity.for("Ben Stein")` <- this exists; there are examples of doing it both ways.

Will think on the other comments. I like that VNode syntax a bit better than what I had in mind. Will definitely seek wider feedback!

LizR3 replied:
So been thinking about this. I don't like the `Vnode.for` - it's another class and syntax that a dev has to magically know about; I feel like once you do `FakeCorda()` everything else should be able to be worked out from there. Returning a VNode would be better IMO.

But, see my post on the Platform channel. When we invoke a node, we don't do it through a node endpoint. We do it through the same endpoint that we used to upload the CPI and create the node. Am I wrong in that understanding? Is there an alternative mechanism that I should know about?

At the moment, "start a flow for X" feels _exactly_ like what I'm doing; the URL is `/flow/{holdingidentityshorthash}`, not `/{holdingidentityshorthash}/flow`, and it uses the same host / port as any other call.

driessamyn replied:
I agree that our REST API is not correct. It should be `https://localhost:8888/api/v1/vnode/$HOLDING_ID/flow/$FLOW_NAME`

Some debate about this in https://r3-cev.atlassian.net/wiki/spaces/CB/pages/3852435478/February+2022+C5+Example+API+Design but I don't think this was actioned.

LizR3 replied:
Agree that would be better syntax.

It still goes through the same RPC endpoint, and as @mnesbit said in channel, permissions can end up limiting which nodes users can talk to, so pulling the node out of FakeCorda still feels like the wrong thing to do.

I have however tweaked the language a bit and am now returning a VirtualNodeInfo - does this work better?

driessamyn replied:
I'm not sure I understand the "endpoint" argument.
Alice and bob may be on completely different clusters and therefore you would invoke flows on completely different URLs. A VNode us a resource on a cluster, and a sub resource is a flow.

To me `corda.callFlow` doesn't feel natural.
What is "Corda" in your context? I assumed it was a test representation of the network (hence you deploy vnodes to it), so you don't really call a flow on "Corda" (the network) for a particular vnode, you _invoke_ or _start_ a flow on a VNode and therefore you must know how to address that vnode.

It's all scemantics, but I think it's important it makes sense from how a cordapp developer sees the network/node/identity, not from the perspective of the physical architecture or topology.
In many ways, the way MockNetwork is structured still makes a lot of sense (https://docs.r3.com/en/platform/corda/4.6/open-source/api-testing.html).

driessamyn reviewed (commented):
Just my 5c.
But this could do with a bit of a design proposal (https://github.com/corda/platform-eng-design/) and feedback/input from a wider group

LizR3 committed:
CORE-6140: FakeCorda returning VirtualNodeInfo

In real Corda, uploading or forcing upload of a CPI returns information about the virtual node created, particularly the short id. This change mimics that return of info and uses it in subsequent flow invocations.

LizR3 committed:
CORE-6140: Virtual nodes run flows

driessamyn reviewed (commented):
Looks reasonable to me, but worth going through with somebody who writes CorDapps for a living ;)

mattbradburyr3 commented:
Hi Liz, It looks good at a cursor look through, so I'll approve the PR. However, I need to get the chance to use the framework as a whole to comment on its use-ability.

mattbradburyr3 reviewed (approved):
Change to using VNs looks good. However I need to do a wider review on use-ability
